### ISSUE TITLE
image_picker logs Exception when trying to select a PNG image using maxWidth and maxHeight

### ISSUE THREAD

**Issue by sceee** [flutter/flutter#64999]
## Steps to Reproduce

1. Use `image_picker 0.6.7+7` plugin
2. Pick a .png image from locally or Google Drive using the following code
```
    final picker = ImagePicker();
    final pickedFile = await picker.getImage(
      source: imgSource,
      maxWidth: 2048.0,
      maxHeight: 2048.0,
    );

    if (pickedFile != null) {
      final imageBytes = await pickedFile.readAsBytes();
    }
```

**Expected results:** Image is picked correctly, no Exception is logged. Docs say that `Compression is only supportted for certain image types such as JPEG. If compression is not supported for the image that is picked, an warning message will be logged.` but as I'm not using the compression setting, I am wondering why the following exception is logged.

**Actual results:** Image is picked but Exception is also logged.

<details>
  <summary>
Invalid image: ExifInterface got an unsupported image format file(ExifInterface supports JPEG and some RAW image formats only) or a corrupted JPEG file to ExifInterface.
</summary>

```
W/ExifInterface(21725): Invalid image: ExifInterface got an unsupported image format file(ExifInterface supports JPEG and some RAW image formats only) or a corrupted JPEG file to ExifInterface.
W/ExifInterface(21725): java.io.IOException: Invalid byte order: ffff8950
W/ExifInterface(21725): 	at android.media.ExifInterface.readByteOrder(ExifInterface.java:3121)
W/ExifInterface(21725): 	at android.media.ExifInterface.isOrfFormat(ExifInterface.java:2437)
W/ExifInterface(21725): 	at android.media.ExifInterface.getMimeType(ExifInterface.java:2315)
W/ExifInterface(21725): 	at android.media.ExifInterface.loadAttributes(ExifInterface.java:1753)
W/ExifInterface(21725): 	at android.media.ExifInterface.initForFilename(ExifInterface.java:2297)
W/ExifInterface(21725): 	at android.media.ExifInterface.<init>(ExifInterface.java:1384)
W/ExifInterface(21725): 	at io.flutter.plugins.imagepicker.ExifDataCopier.copyExif(ExifDataCopier.java:15)
W/ExifInterface(21725): 	at io.flutter.plugins.imagepicker.ImageResizer.copyExif(ImageResizer.java:120)
W/ExifInterface(21725): 	at io.flutter.plugins.imagepicker.ImageResizer.resizeImageIfNeeded(ImageResizer.java:49)
W/ExifInterface(21725): 	at io.flutter.plugins.imagepicker.ImagePickerDelegate.handleImageResult(ImagePickerDelegate.java:556)
W/ExifInterface(21725): 	at io.flutter.plugins.imagepicker.ImagePickerDelegate.handleChooseImageResult(ImagePickerDelegate.java:492)
W/ExifInterface(21725): 	at io.flutter.plugins.imagepicker.ImagePickerDelegate.onActivityResult(ImagePickerDelegate.java:471)
W/ExifInterface(21725): 	at io.flutter.embedding.engine.FlutterEnginePluginRegistry$FlutterEngineActivityPluginBinding.onActivityResult(FlutterEnginePluginRegistry.java:691)
W/ExifInterface(21725): 	at io.flutter.embedding.engine.FlutterEnginePluginRegistry.onActivityResult(FlutterEnginePluginRegistry.java:378)
W/ExifInterface(21725): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult(FlutterActivityAndFragmentDelegate.java:619)
W/ExifInterface(21725): 	at io.flutter.embedding.android.FlutterActivity.onActivityResult(FlutterActivity.java:584)
W/ExifInterface(21725): 	at android.app.Activity.dispatchActivityResult(Activity.java:8110)
W/ExifInterface(21725): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4838)
W/ExifInterface(21725): 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4886)
W/ExifInterface(21725): 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
W/ExifInterface(21725): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
W/ExifInterface(21725): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
W/ExifInterface(21725): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
W/ExifInterface(21725): 	at android.os.Handler.dispatchMessage(Handler.java:107)
W/ExifInterface(21725): 	at android.os.Looper.loop(Looper.java:214)
W/ExifInterface(21725): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
W/ExifInterface(21725): 	at java.lang.reflect.Method.invoke(Native Method)
W/ExifInterface(21725): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
W/ExifInterface(21725): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```
</details>

```
[√] Flutter (Channel stable, 1.20.2, on Microsoft Windows [Version 10.0.18363.1016], locale de-DE)
    • Flutter version 1.20.2 at C:\Dev\flutter
    • Framework revision bbfbf1770c (3 weeks ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1

 
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\...\AppData\Local\Android\sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.48.2)
    • VS Code at C:\Users\...\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.13.2

[√] Connected device (1 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)

• No issues found!
```

**Comment by pedromassangocode** [comment]
Hi @sceee 
Can you please provide your `flutter doctor -v`?
Thank you

**Comment by sceee** [comment]
@pedromassangocode sorry, it was already included but in the collapsed part. Updated it now.

**Comment by pedromassangocode** [comment]
I was not able to reproduce this on latest version of the `image_picker` package.

Tried with the image in https://www.pngfind.com/pngs/m/45-451791_free-png-download-lionel-messi-png-images-background.png

Does the code bellow and the image above reproduces the issue?
Can you share the image you are using to reproduce this issue?

<details>
<summary> sample code</summary>

```
// Copyright 2019 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// ignore_for_file: public_member_api_docs

import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Image Picker Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  void _onImageButtonPressed(ImageSource source, {BuildContext context}) async {
    try {
      final data = await ImagePicker().getImage(
        source: source,
        maxWidth: 2048.0,
        maxHeight: 2048.0,
      );

      if (data != null) {
        final imageBytes = await data.readAsBytes();
        print('Image file picked:');
        print('Image file picked:');
        print('Image file picked:');
      }
    } catch (e) {
      print(e);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Text(
        'You have not yet picked an image.',
        textAlign: TextAlign.center,
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _onImageButtonPressed(ImageSource.gallery, context: context);
        },
        heroTag: 'image0',
        tooltip: 'Pick Image from gallery',
        child: const Icon(Icons.photo_library),
      ),
    );
  }
}

```
</details>

<details>
<summary> Logs</summary>

```
Performing hot reload...
Syncing files to device sdk gphone x86 arm...
Reloaded 1 of 527 libraries in 435ms.
E/Surface ( 6881): getSlotFromBufferLocked: unknown buffer: 0x0
E/Surface ( 6881): getSlotFromBufferLocked: unknown buffer: 0xeccf4790
I/evercode.triag( 6881): NativeAlloc concurrent copying GC freed 8382(674KB) AllocSpace objects, 27(1852KB) LOS objects, 49% free, 2301KB/4603KB, paused 135us total 162.485ms
I/flutter ( 6881): Image file picked:
I/chatty  ( 6881): uid=10154(com.nevercode.triage) 1.ui identical 1 line
I/flutter ( 6881): Image file picked:
```
</details>

<details>
<summary> flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.20.2, on Mac OS X 10.15.6 19G2021, locale en)
    • Flutter version 1.20.2 at /Users/pedromassango/dev/SDKs/flutter_stable
    • Framework revision bbfbf1770c (3 weeks ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1

 
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor
      --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 48.1.4
    • Dart plugin version 202.7206

[!] VS Code (version 1.48.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (1 available)
    • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86 • Android 11
      (API 30) (emulator)

! Doctor found issues in 3 categories.
```
</details>

**Comment by sceee** [comment]
@pedromassangocode with your provided image, I cannot reproduce the issue.

I can unfortunately not share the original image but I created another test image that can reproduce it:
![black](https://user-images.githubusercontent.com/11340487/92249512-86c3ba00-eeca-11ea-9eac-d1cd95716096.png)

It's pure black, freshly created in paint.net 4.2.13, saved as PNG with auto-detected bit-depth setting. Please find the save settings here (sorry for it being german but you can probably locate it):
![image](https://user-images.githubusercontent.com/11340487/92249660-bbd00c80-eeca-11ea-9ad5-afd69c3b353e.png)

I can of course not rule out that paint.net produces incorrect images but I would be surprised if it does.

**Comment by pedromassangocode** [comment]
Hi @sceee 
Are you able to reproduce with the code bellow?

<details>
<summary> sample code</summary>

```
// Copyright 2019 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// ignore_for_file: public_member_api_docs

import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Image Picker Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  void _onImageButtonPressed(ImageSource source, {BuildContext context}) async {
    try {
      final data = await ImagePicker().getImage(
        source: source,
        maxWidth: 2048.0,
        maxHeight: 2048.0,
      );

      if (data != null) {
        final imageBytes = await data.readAsBytes();
        print('Image file picked:');
        print('Image file picked:');
        print('Image file picked:');
      }
    } catch (e) {
      print(e);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Text(
        'You have not yet picked an image.',
        textAlign: TextAlign.center,
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _onImageButtonPressed(ImageSource.gallery, context: context);
        },
        heroTag: 'image0',
        tooltip: 'Pick Image from gallery',
        child: const Icon(Icons.photo_library),
      ),
    );
  }
}

```
</details>

**Comment by sceee** [comment]
@pedromassangocode yes, it's also reproducible using the image I provided you with your code.

**Comment by pedromassangocode** [comment]
I was able to reproduce this with the latest version of the `image_picker: ^0.6.7+7` package with Emulator API 29. This issue seems to reproduce only on Android 10 (API 29), because I was not able to reproduce with Android 11 emulator (API 30).

Important to notice that the image is picked successfully, but it looks like the native code tries to convert it and it throws the exception.

<details><summary>Logs</summary>

```
W/ExifInterface( 5618): Invalid image: ExifInterface got an unsupported image format file(ExifInterface supports JPEG and some RAW image formats only) or a corrupted JPEG file to ExifInterface.
W/ExifInterface( 5618): java.io.IOException: Invalid byte order: ffff8950
W/ExifInterface( 5618): 	at android.media.ExifInterface.readByteOrder(ExifInterface.java:3121)
W/ExifInterface( 5618): 	at android.media.ExifInterface.isOrfFormat(ExifInterface.java:2437)
W/ExifInterface( 5618): 	at android.media.ExifInterface.getMimeType(ExifInterface.java:2315)
W/ExifInterface( 5618): 	at android.media.ExifInterface.loadAttributes(ExifInterface.java:1753)
W/ExifInterface( 5618): 	at android.media.ExifInterface.initForFilename(ExifInterface.java:2297)
W/ExifInterface( 5618): 	at android.media.ExifInterface.<init>(ExifInterface.java:1384)
W/ExifInterface( 5618): 	at io.flutter.plugins.imagepicker.ExifDataCopier.copyExif(ExifDataCopier.java:15)
W/ExifInterface( 5618): 	at io.flutter.plugins.imagepicker.ImageResizer.copyExif(ImageResizer.java:120)
W/ExifInterface( 5618): 	at io.flutter.plugins.imagepicker.ImageResizer.resizeImageIfNeeded(ImageResizer.java:49)
W/ExifInterface( 5618): 	at io.flutter.plugins.imagepicker.ImagePickerDelegate.handleImageResult(ImagePickerDelegate.java:556)
W/ExifInterface( 5618): 	at io.flutter.plugins.imagepicker.ImagePickerDelegate.handleChooseImageResult(ImagePickerDelegate.java:492)
W/ExifInterface( 5618): 	at io.flutter.plugins.imagepicker.ImagePickerDelegate.onActivityResult(ImagePickerDelegate.java:471)
W/ExifInterface( 5618): 	at io.flutter.embedding.engine.FlutterEnginePluginRegistry$FlutterEngineActivityPluginBinding.onActivityResult(FlutterEnginePluginRegistry.java:691)
W/ExifInterface( 5618): 	at io.flutter.embedding.engine.FlutterEnginePluginRegistry.onActivityResult(FlutterEnginePluginRegistry.java:378)
W/ExifInterface( 5618): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult(FlutterActivityAndFragmentDelegate.java:619)
W/ExifInterface( 5618): 	at io.flutter.embedding.android.FlutterActivity.onActivityResult(FlutterActivity.java:584)
W/ExifInterface( 5618): 	at android.app.Activity.dispatchActivityResult(Activity.java:8110)
W/ExifInterface( 5618): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4838)
W/ExifInterface( 5618): 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4886)
W/ExifInterface( 5618): 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
W/ExifInterface( 5618): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
W/ExifInterface( 5618): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
W/ExifInterface( 5618): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
W/ExifInterface( 5618): 	at android.os.Handler.dispatchMessage(Handler.java:107)
W/ExifInterface( 5618): 	at android.os.Looper.loop(Looper.java:214)
W/ExifInterface( 5618): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
W/ExifInterface( 5618): 	at java.lang.reflect.Method.invoke(Native Method)
W/ExifInterface( 5618): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
W/ExifInterface( 5618): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
D/EGL_emulation( 5618): eglMakeCurrent: 0xe2befa00: ver 3 0 (tinfo 0xd75523d0)
D/EGL_emulation( 5618): eglMakeCurrent: 0xe2bef100: ver 3 0 (tinfo 0xd75512c0)
D/EGL_emulation( 5618): eglMakeCurrent: 0xe2befa00: ver 3 0 (tinfo 0xd75523d0)
I/flutter ( 5618): Image file picked:
I/chatty  ( 5618): uid=10137(com.nevercode.triage) 1.ui identical 1 line
I/flutter ( 5618): Image file picked:

```
</details>

<details><summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.22.0-10.0.pre.87, on Mac OS X 10.15.6 19G2021, locale en)
    • Flutter version 1.22.0-10.0.pre.87 at /Users/pedromassango/dev/SDKs/flutter_master
    • Framework revision 4732a214a7 (2 days ago), 2020-09-04 21:05:02 -0400
    • Engine revision ac8b9c4c52
    • Dart version 2.10.0 (build 2.10.0-86.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.7)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.7, Build version 11E801a
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 48.1.4
    • Dart plugin version 202.7206

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (4 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • macOS (desktop)                    • macos         • darwin-x64     • Mac OS X 10.15.6 19G2021
    • Web Server (web)                   • web-server    • web-javascript • Flutter Tools
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 85.0.4183.83

! Doctor found issues in 1 category.
```
</details>

**Comment by valconetti** [comment]
Hi @sceee, 

It's correct that it doesn't work with PNG, it states that in [the documentation](https://pub.dev/documentation/image_picker/latest/image_picker/ImagePicker/getImage.html).

> Compression is only supportted for certain image types such as JPEG. If compression is not supported for the image that is picked, an warning message will be logged.

 The reason that you don't get the error message with the pic of Messi, is because it's in reality a JPEG when downloaded. You can check this by opening it [here](https://hexed.it/) after downloading it and checking the first bytes. JPEG start with `FF D8 FF` and a PNG starts with `89 50 4E 47 0D 0A 1A 0A`. If you convert it to a PNG and try opening it with the `image_picker` again, you should get the same error message. 

What happens is that `copyExif` copies the meta data of the picture and puts it in a different frame, using [`ExifInterface`](https://developer.android.com/reference/android/media/ExifInterface). `ExifInterface` is an [Android object class](https://developer.android.com/reference/android/media/ExifInterface), which is:

> ... for reading and writing Exif tags in a JPEG file or a RAW image file.

If you run it on API 29, it goes to the `ExifInterface`, checks whether the file is a JPEG or a RAW image file and gives the error of the picture not being supported. The error you're getting is in reality just a warning log, of which the problem is handled internally. The app continues to run, because the the error is not a crash but a warning. Running it on API 30 it passes the check, since it uses the `AndroidX` library of that class which also [works with PNG images.](https://developer.android.com/reference/androidx/exifinterface/media/ExifInterface):

> Supported for reading: JPEG, PNG, WebP, HEIF, DNG, CR2, NEF, NRW, ARW, RW2, ORF, PEF, SRW, RAF.
> Supported` for writing: JPEG, PNG, WebP.

Because of that I'm changing the `ExifInterface` to the `AndroidX` version of the library for the other API's, too. After that you shouldn't get the error anymore.

**Comment by sceee** [comment]
@valconetti thanks - as mentioned in the first post, I understood that is is only a logged exception, not actually causing harm. My only point was it might confuse people (as it did me in the first place).

Anyways, thanks!

**Comment by mvanbeusekom** [comment]
Closing this issue as a fix was provided in version 0.6.7+9.

**Comment by diegoveloper** [comment]
I'm getting the same warning using the latest version.

```
D/ImageResizer( 3230): image_picker: compressing is not supported for type PNG. Returning the image with original quality

```

Flutter doctor:
```
 Flutter (Channel stable, 1.22.6, on Mac OS X 10.15.7 19H2 darwin-x64, locale en-US)
```

Plugin version:
```
  image_picker:
    dependency: "direct main"
    description:
      name: image_picker
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.6.7+22"
```

Tested on Android , Pixel 5 running Android 11.0
